### PR TITLE
Add support for asynchronous typescript mocks

### PR DESCRIPTION
**Pull request by DSoko2**
# Description

This PR implements the feature request #13049.

Pulumi's mocking feature for TypeScript is configured by calling `pulumi.runtime.setMocks(...)`. It only supports synchronous mock implementations. I.e., the functions call and `newResource`, which are provided as an object of type Mocks, must be synchronous. This prevents using any asynchronous function or APIs to generate or retrieve values to return because they cannot be awaited in a synchronous TS function.

This PR extends the `Mocks` interface with support for asynchronous mock implementations. The changes are backward compatible. In contrast to the implementation sketched in [#13049](https://github.com/pulumi/pulumi/issues/13049), this version is more elegant, requires less code, and is still type-safe.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by RobbieMcKinstry** [comment]
bors try

File: sdk/nodejs/runtime/mocks.ts [hunk 1]
```diff
@@ -75,7 +84,12 @@ export interface MockCallArgs {
 }
 
 /**
- * Mocks is an abstract class that allows subclasses to replace operations normally implemented by the Pulumi engine with
+ * MockCallResult is the expect result of a call Mock.
```
Comment (Frassle) [inline sdk/nodejs/runtime/mocks.ts:87]:
```suggestion
 * MockCallResult is the result of a call Mock.
```

File: sdk/nodejs/runtime/mocks.ts [hunk 2]
```diff
@@ -55,7 +55,16 @@ export interface MockResourceArgs {
 }
 
 /**
- * MockResourceArgs is used to construct call Mock
+ * MockResourceResult is the expect result of a newResource Mock, returning a physical identifier and the output properties
```
Comment (Frassle) [inline sdk/nodejs/runtime/mocks.ts:58]:
```suggestion
 * MockResourceResult is the result of a newResource Mock, returning a physical identifier and the output properties
```

**Review by Frassle — COMMENTED** [review]
Couple of documentation changes suggested, but otherwise this looks fine to me.

**Review by Frassle — APPROVED** [review]
Thanks this looks reasonable to me

bors merge

**Pull request merged**